### ISSUE TITLE
Reflection bugs

### ISSUE THREAD

**Issue by enumag** [php-ds/ext-ds#119]
Currently reflection on some `Ds\*` classes doesn't return the expected values which is causing issues with tools such as [PHPStan](https://github.com/phpstan/phpstan).

### Missing `Ds\Pair` properties

```php
$reflection = new ReflectionClass(\Ds\Pair::class);

echo count($reflection->getProperties());
```

expected result: 2 (Pair has $key and $value properties)

actual result: 0

These properties are missing in the [documentation](http://php.net/manual/en/class.ds-pair.php) as well (bug report [here](https://bugs.php.net/bug.php?id=76687)).

EDIT: Ds\Pair values are also invisible when debuging with Xdebug. This might be the cause.

### Nullability of `Map::first()`

```php
$reflection = new ReflectionClass(\Ds\Map::class);

echo $reflection->getMethod('first')->getReturnType()->allowsNull() ? 'true' : 'false';
```

expected result: false (the method throws an exception on empty map and otherwise returns a Pair)

actual result: true

EDIT: `Map::last()` and `Map::skip()` both have the same problem.

**Comment by enumag** [comment]
Another bug discovered:

### Nullability of `Map::values()`

```php
$reflection = new ReflectionClass(\Ds\Map::class);

echo $reflection->getMethod('values')->getReturnType()->allowsNull() ? 'true' : 'false';
```

expected result: false (return type is `Ds\Sequence`, not `?Ds\Sequence` - see [documentation](http://php.net/manual/en/ds-map.values.php))

actual result: true

**Comment by enumag** [comment]
Looks like this is just the tip of an iceberg, there are a lot more methods in this extension with incorrect reflection types. I have a tool to generate a functionMap for phpstan which I then fixed manually and I could use it to give you a more complete list what's wrong. I won't do it until I see some activity here since it would take me quite a bit of time and it's possible that you don't even need it.

**Comment by rtheunissen** [comment]
@enumag are you able to use a master build? This should now be significantly better than 1.6, due to this change here: https://github.com/php-ds/extension/commit/21cbf36703aed19f9a6c9c429e5754cb3687905c

**Comment by rtheunissen** [comment]
I'll prepare an update regardless.

**Comment by rtheunissen** [comment]
Yeah all the nullable return types should be fixed, but I haven't fixed the missing Pair properties yet. There's a quick-hack fix for it but I'd rather do it properly. They are accessed magically but should be defined properties.

**Comment by rtheunissen** [comment]
> I installed the extension using apt-get

I'm not sure who maintains this.. 🤔

**Comment by rtheunissen** [comment]
What version are you using?

**Comment by rtheunissen** [comment]
These should now all be fixed as at 1.2.7 :tada:
